File: languages/python/bloock/entity/integrity/proof.py
from __future__ import annotations

from typing import List

import bloock._bridge.proto.integrity_entities_pb2 as proto
from bloock.entity.integrity.proof_anchor import ProofAnchor


class Proof:
    """
    Represents a proof, including leaves, nodes, depth, bitmap, and anchor information.
    """
    def __init__(
            self,
            leaves: List[str],
            nodes: List[str],
            depth: str,
            bitmap: str,
            anchor: ProofAnchor,
    ) -> None:
        """
        Constructs a Proof object with the specified parameters.
        :type anchor: object
        :type bitmap: object
        :type depth: object
        :type nodes: object
        :type leaves: object
        :rtype: object
        """
        self.leaves = leaves
        self.nodes = nodes
        self.depth = depth
        self.bitmap = bitmap
        self.anchor = anchor

    @staticmethod
    def from_proto(proof: proto.Proof) -> Proof:
        return Proof(
            leaves=list(proof.leaves),
            nodes=list(proof.nodes),
            depth=proof.depth,
            bitmap=proof.bitmap,
            anchor=ProofAnchor.from_proto(proof.anchor),
        )

    def to_proto(self) -> proto.Proof:
        return proto.Proof(
            leaves=self.leaves,
            nodes=self.nodes,
            depth=self.depth,
            bitmap=self.bitmap,
            anchor=self.anchor.to_proto(),
        )

File: languages/php/Bloock/Entity/Integrity/Proof.php
<?php

namespace Bloock\Entity\Integrity;

/**
 * Represents a proof, including leaves, nodes, depth, bitmap, and anchor information.
 */
class Proof
{
    private array $leaves;
    private array $nodes;
    private string $depth;
    private string $bitmap;
    private ProofAnchor $anchor;

    /**
     * Constructs a Proof object with the specified parameters.
     * @param array $leaves
     * @param array $nodes
     * @param string $depth
     * @param string $bitmap
     * @param ProofAnchor $anchor
     */
    public function __construct(array $leaves, array $nodes, string $depth, string $bitmap, ProofAnchor $anchor)
    {
        $this->leaves = $leaves;
        $this->nodes = $nodes;
        $this->depth = $depth;
        $this->bitmap = $bitmap;
        $this->anchor = $anchor;
    }

    public static function fromProto(\Bloock\Proof $proof): Proof
    {
        $leaves = [];
        foreach ($proof->getLeaves()->getIterator() as $item) {
            $leaves[] = $item;
        }

        $nodes = [];
        foreach ($proof->getNodes()->getIterator() as $item) {
            $nodes[] = $item;
        }
        return new Proof($leaves, $nodes, $proof->getDepth(), $proof->getBitmap(), ProofAnchor::fromProto($proof->getAnchor()));
    }

    /**
     * Gets the leaves of the proof.
     * @return array
     */
    public function getLeaves(): array
    {
        return $this->leaves;
    }

    /**
     * Gets the nodes of the proof.
     * @return array
     */
    public function getNodes(): array
    {
        return $this->nodes;
    }

    /**
     * Gets the depth of the proof.
     * @return string
     */
    public function getDepth(): string
    {
        return $this->depth;
    }

    /**
     * Gets the bitmap of the proof.
     * @return string
     */
    public function getBitmap(): string
    {
        return $this->bitmap;
    }

    /**
     * Gets the anchor of the proof.
     * @return ProofAnchor
     */
    public function getAnchor(): ProofAnchor
    {
        return $this->anchor;
    }

    public function toProto(): \Bloock\Proof
    {
        $p = new \Bloock\Proof();
        $p->setLeaves($this->leaves);
        $p->setNodes($this->nodes);
        $p->setDepth($this->depth);
        $p->setBitmap($this->bitmap);
        $p->setAnchor($this->anchor->toProto());
        return $p;
    }
}